File: Yixuan/18_01_17/search_a_2D_matrix_II.java
public class Solution {
    /**
     * @param matrix: A list of lists of integers
     * @param: A number you want to search in the matrix
     * @return: An integer indicate the occurrence of target in the given matrix
     */
    public int searchMatrix(int[][] matrix, int target) {
      if (matrix.length == 0 || matrix == null) {
         return 0;
      }
      if (matrix[0].length == 0 || matrix[0] == null) {
         return 0;
      }
      int row = matrix.length;
      int column = matrix[0].length;
      int x = row - 1;
      int y = 0;
      int occurrence = 0;
      while (x >= 0 && y <= column - 1) {
         int curr_num = matrix[x][y];
         if (target == curr_num) {
            occurrence++;
            x--;
            y++;
         } else if (target > curr_num) {
            y++;
         } else {
            x--;
         }
      }
      return occurrence;
    }
}

File: Boxuan/16_01_18/search_a_2d_matrix_II.py
class Solution:
    """
    @param matrix: An list of lists of integers
    @param target: An integer you want to search in matrix
    @return: An integer indicates the total occurrence of target in the given matrix
    """
    def searchMatrix(self, matrix, target):
        if not matrix:
            return 0
        row, col = len(matrix) - 1, 0
        counter = 0
        while col < len(matrix[0]) and row >= 0:
            if matrix[row][col] == target:
                row -= 1
                col += 1
                counter += 1
            elif matrix[row][col] > target:
                row -= 1
            elif matrix[row][col] < target:
                col += 1
        return counter